### PR TITLE
contib: add direnv implementation

### PR DESCRIPTION
**Pull request by blaggacao**
### Without `PRJ_ID`:
```console
$ direnv reload # initial
[ ... ]
direnv: Add to .git/info/exclude: /.config
direnv: Add to .git/info/exclude: /.cache
direnv: Add to .git/info/exclude: /.local/share
direnv: Add to .git/info/exclude: /.local/state
direnv: Add to .git/info/exclude: /.run
direnv: PRJ_ROOT:        /home/blaggacao/src/github.com/paisano-nix/tui
direnv: PRJ_ID:          none
direnv: PRJ_CONFIG_HOME: .config
direnv: PRJ_RUNTIME_DIR: .run
direnv: PRJ_CACHE_HOME:  .cache
direnv: PRJ_DATA_HOME:   .local/share
direnv: PRJ_STATE_HOME:  .local/state
direnv: PRJ_PATH:        .local/bin
[ ... ]
```
```console
$ direnv reload # subsequent
[ ... ]
direnv: PRJ_ROOT:        /home/blaggacao/src/github.com/paisano-nix/tui
direnv: PRJ_ID:          none
direnv: PRJ_CONFIG_HOME: .config
direnv: PRJ_RUNTIME_DIR: .run
direnv: PRJ_CACHE_HOME:  .cache
direnv: PRJ_DATA_HOME:   .local/share
direnv: PRJ_STATE_HOME:  .local/state
direnv: PRJ_PATH:        .local/bin
[ ... ]
```
### With `PRJ_ID`:
```console
uuidgen > .config/prj_id
```
```console
$ direnv reload # initial
[ ... ]
direnv: Add to .git/info/exclude: /.config
direnv: Add to .git/info/exclude: /.run
direnv: PRJ_ROOT:        /home/blaggacao/src/github.com/paisano-nix/tui
direnv: PRJ_ID:          44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_CONFIG_HOME: .config
direnv: PRJ_RUNTIME_DIR: .run
direnv: PRJ_CACHE_HOME:  /home/blaggacao/.cache/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_DATA_HOME:   /home/blaggacao/.local/share/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_STATE_HOME:  /home/blaggacao/.local/state/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_PATH:        /home/blaggacao/.local/bin/prj/44850561-5378-499b-9e95-8549b4bc8596
[ ... ]
```
```console
$ direnv reload # subsequent
[ ... ]
direnv: PRJ_ROOT:        /home/blaggacao/src/github.com/paisano-nix/tui
direnv: PRJ_ID:          44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_CONFIG_HOME: .config
direnv: PRJ_RUNTIME_DIR: .run
direnv: PRJ_CACHE_HOME:  /home/blaggacao/.cache/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_DATA_HOME:   /home/blaggacao/.local/share/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_STATE_HOME:  /home/blaggacao/.local/state/prj/44850561-5378-499b-9e95-8549b4bc8596
direnv: PRJ_PATH:        /home/blaggacao/.local/bin/prj/44850561-5378-499b-9e95-8549b4bc8596
[ ... ]
```

### Silence, please ...
```console
$ export DIRENV_PRJ_SILENCE=true;
```
```console
$ direnv reload # initial
[ ... ]
direnv: PRJ_CONFIG_HOME: .config
direnv: PRJ_RUNTIME_DIR: .run
# more if PRJ_ID is not set
[ ... ]
```

```console
$ direnv reload # subsequent
[ ... ]
```

### CODE REVIEW

File: contrib/direnv
Comment (zimbatm) [inline contrib/direnv:70]: Instead, we could rename `.local/share` to `${PRJ_ROOT}/.data`. I'm not sure why the XDG spec picked `.local/share`, maybe because of NFS mounts?

File: contrib/direnv
Comment (zimbatm) [inline contrib/direnv:97]: I know it's convenient, but projects should probably support this explicitly by setting up their .gitignore accordingly?

File: contrib/direnv
Comment (zimbatm) [inline contrib/direnv:106]: I'm not sure what's the use-case for PRJ_PATH?

File: contrib/direnv
Comment (zimbatm) [inline contrib/direnv:111]: What is the difference between PRJ_DATA_HOME, PRJ_STATE_HOME and PRJ_RUNTIME_DIR?

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:119]: Yes, exactly. And educational routine (and once educated `echo > DIRENV_PRJ_SILENCE=please > .envrc.local`)

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:24]: Yes, this is a "deviaton" from the XDG convention.
This is mainly useful for dev services, which can be configured to use it and the user will be able to find it in the same place.

For example `mariadb` by default can't be killed with  `CTRL+D` so having the `pid` file at hand may end up being handy for some extra whichcraft. Or another thing could be project-local service sockets....

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:26]: This is really intended to be `GIT_DIR` (not root). I wasn't confident if `$(git rev-parse --show-toplevel)/.git` would fulfill the same contract under all circumstances. But yeah, two lines afterwards, I still consume that exact same contract LOL.

I just so noticed.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:70]: Agreed! That makes actually more intuitive sense!

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:106]: Potentially (big one!) `nix` or `cargo` could (be configured to) place their files there.

That makes a handy contract that is already in the `PATH` (`PATH_add`) regardless of the tool's choice.

As the rest of this spec, it best works if _tool_ start to comply.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:111]: According to `XDG`, `DATA` is for potentially sharable data that is useful independent of the context of the concrete user environment. One example that comes to my mind would be blockchain state. A way to think about it would be a hypothetical user that works on mac and linux interchangeably but shares projects on a NFS. `DATA` should _not_ hold arch-specific data.

`STATE`, however, could be a stateful application file (these are rare these days), that don't immediately make sense to share, either because they are arch-specific or their lifetime is bound to the host machine. A better use in most cases would be `/tmp`, but still, maybe an application stores last startup configuration to resume on. The Paisano TUI/CLI is recording the last executed action there in order to facilitate manual debugging but also because the invocation sequence requires a file proxy at some point.

Comment (blaggacao) [reply contrib/direnv:111]: `PRJ_RUNTIME_DIR` is usually just pids and sockets, maybe a runtime wal or such.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:97]: Maybe,  yes. That bit of convenience can be optional.

My reasoning, otoh, was that I couldn't think of a legit use case to commit these files and otherwise [out of my own experience] people tend to have allergic reactions if files pop up in their diffs which they didn't hand craft themselves.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:7]: Yes, this can be too ambiguous. And I already sort of thought about this problem, but then chose not to address it on the first iteration. But it needs to be addressed before merge.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:34]: This is level 2, indeed. And we should make it configurable.

However, `direnv` is rightly used to prepare the environment. Imagine writhing scripts. You'd like to consume level 2 of this spec (folders created) or else you can get screwed if you ever forget a verbose `mkdir -p $PRJ_*` anywhere at the consuming site.

File: contrib/direnv
Comment (cscutcher) [inline contrib/direnv:74]: I think this comment is perhaps incorrect. 

`GIT_DIR` as set by `git rev-parse --absolute-git-dir` **could** be set, but we could be outside a worktree. For example, if we were in a "bare" git repository or within the `.git` git subdir within the main worktree then `GIT_DIR` would still be set.

To be fair, for the inverse, i.e. if it's **not* set, then we're definitely **not in** work tree. 

Putting aside the reservations below about doing this at all, if you were to keep it, I'd advise something like;

 ```suggestion
	# Git ignore all `PRJ_*` in the worktree for this repository.
	# Skip this if we're not in a git repository
	# i.e. if we're not in a git repository, GIT_DIR is not a file
```

To get some of this straight in my head for this comment and others, I wrote this [little hacky demo script](https://gist.github.com/cscutcher/e55c6351c7dd60ef7743186ad89eccb8#file-demo-sh) with output that I hope demonstrates;

```console
$ cd example_repo && echo /tmp
/tmp



$ # Example of different git values within main worktree...
$ git worktree add ../example_repo_worktree
Preparing worktree (checking out 'example_repo_worktree')
HEAD is now at 03370b6 Initial commit
$ git rev-parse --show-toplevel
/tmp/example_repo
$ git rev-parse --absolute-git-dir
/tmp/example_repo/.git
$ git rev-parse --is-inside-work-tree
true
$ git rev-parse --is-inside-git-dir
false



$ cd .git
$ git rev-parse --show-toplevel
fatal: this operation must be run in a work tree
$ git rev-parse --absolute-git-dir
/tmp/example_repo/.git
$ git rev-parse --is-inside-work-tree
false
$ git rev-parse --is-inside-git-dir
true



$ # Example of different git values within a linked worktree...
$ cd ../../example_repo_worktree
$ git rev-parse --show-toplevel
/tmp/example_repo_worktree
$ git rev-parse --absolute-git-dir
/tmp/example_repo/.git/worktrees/example_repo_worktree
$ git rev-parse --is-inside-work-tree
true
$ git rev-parse --is-inside-git-dir
false



$ # Just for completeness, the output of these when outside a git repo
$ cd /tmp
$ git rev-parse --show-toplevel
fatal: not a git repository (or any of the parent directories): .git
$ git rev-parse --absolute-git-dir
fatal: not a git repository (or any of the parent directories): .git
$ git rev-parse --is-inside-work-tree
fatal: not a git repository (or any of the parent directories): .git
$ git rev-parse --is-inside-git-dir
fatal: not a git repository (or any of the parent directories): .git
```

File: contrib/direnv
Comment (cscutcher) [inline contrib/direnv:4]: Minor thing, just to make sure I'm on the right page, is this just for if we can't detect this via Git?

If so, perhaps worth making this more explicit with;

```suggestion
# Discover PRJ_ROOT if not in a git repository.
# Look for the .config folder in the current directory.
find_prj_root_without_git() (
```

you might even want a simple `find_prj_root_with_git` for parity.

**Review by cscutcher — COMMENTED** [review]
Stopping by mostly after discovering (I think) some issues with how this code behaves in an environment with linked git worktrees (i.e. created via `git worktree add`). 

Hope it's OK if I made some other related suggestions while I was in the area. 

As always, thanks for making the effort and sharing the code!

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:27]: Hm, this points to a worktree, too:
```console
❯ git rev-parse --show-toplevel
/home/blaggacao/src/github.com/input-output-hk/ww
```

While the other command at least has the right repo in the path:
```console
❯ git rev-parse --absolute-git-dir
/home/blaggacao/src/github.com/input-output-hk/wallet-world/.git/worktrees/ww
```

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: dropped in 4b57523d564de284289cc41c38bf0bf6b9c3d5c2 :smile:

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: Removed `PRJ_STATE_HOME` in 7dfd31d4d82c6ebdec54eface81dfb5cce689eab.

But I'm keeping `PRJ_RUNTIME_DIR` for deveshell services use case ("find their pids easily").

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:5]: Addressed in 3ea5e756ca4c228aeb8586eb13fa899e295fdca6

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: Fixed in 3ea5e756ca4c228aeb8586eb13fa899e295fdca6

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:32]: fixed in 3ea5e756ca4c228aeb8586eb13fa899e295fdca6

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: fixed in https://github.com/numtide/prj-spec/pull/4/commits/3ea5e756ca4c228aeb8586eb13fa899e295fdca6

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: @zimbatm would that be ok (keep PRJ_RUNTIME_DIR)? Would you like me to expand more on the use case?

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:86]: @zimbatm Is that use case enough motivated? Or should I provide some more context?

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:40]: @zimbatm I'm not too strongly opinionated about this. Would you still prefer me to remove them?

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:98]: @zimbatm Is the opt-out the right approach or should we do opt-in (which probably would defeat the purpose as nobody would care to discover it in the first place)?

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: addressed in 44d2b5fcb558d9864c290459ad096dba13aaa025.

@zimbatm is that imlementation good enough for our purpose here?

File: contrib/direnv
Comment (cscutcher) [reply contrib/direnv]: > Hm, this points to a worktree, too:
> 
> ```
> ❯ git rev-parse --show-toplevel
> /home/blaggacao/src/github.com/input-output-hk/ww
> ```
> 
> While the other command at least has the right repo in the path:
> 
> ```
> ❯ git rev-parse --absolute-git-dir
> /home/blaggacao/src/github.com/input-output-hk/wallet-world/.git/worktrees/ww
> ```

Sorry if I'm wrong here, but don't you want the former here? As I understand it, your goal is to get the root of the currently checked out git commit? 

I tried to describe the behaviour as I understand it below the break, but it's probably easier to show an example output instead. This should show that with the logic from 3ea5e756ca4c228aeb8586eb1 `PRJ_ROOT` always ends up pointing to the main checkout which would be wrong when working in a worktree.

```console
$ git init /tmp/wallet-world-0
hint: Using 'master' as the name for the initial branch. This default branch name
hint: is subject to change. To configure the initial branch name to use in all
hint: of your new repositories, which will suppress this warning, call:
hint:
hint:   git config --global init.defaultBranch <name>
hint:
hint: Names commonly chosen instead of 'master' are 'main', 'trunk' and
hint: 'development'. The just-created branch can be renamed via this command:
hint:
hint:   git branch -m <name>
Initialized empty Git repository in /tmp/wallet-world-0/.git/
$ cd /tmp/wallet-world-0
$ touch some_file
$ git add some_file
$ git commit -m "Initial commit"
[master (root-commit) 8ef7fd9] Initial commit
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 some_file
$ git show --oneline
8ef7fd9 (HEAD -> master) Initial commit
diff --git a/some_file b/some_file
new file mode 100644
index 0000000..e69de29
$ export GIT_DIR=$(git rev-parse --absolute-git-dir)
$ export PRJ_ROOT="${GIT_DIR%/.git*}"
$ echo $GIT_DIR
/tmp/wallet-world-0/.git
$ echo $PRJ_ROOT
/tmp/wallet-world-0
$ ls $PRJ_ROOT
some_file
$ if [[ -e $PRJ_ROOT/some_new_file ]] ; then echo "on new commit"; else echo "on initial commit"; fi
on initial commit
$ unset GIT_DIR





$ git worktree add /tmp/wallet-world-1
Preparing worktree (new branch 'wallet-world-1')
HEAD is now at 8ef7fd9 Initial commit
$ cd /tmp/wallet-world-1
$ ls
some_file
$ touch some_new_file
$ git add some_new_file
$ git commit -m "New commit"
[wallet-world-1 851809f] New commit
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 some_new_file
$ git show --oneline
851809f (HEAD -> wallet-world-1) New commit
diff --git a/some_new_file b/some_new_file
new file mode 100644
index 0000000..e69de29
$ export GIT_DIR=$(git rev-parse --absolute-git-dir)
$ export PRJ_ROOT="${GIT_DIR%/.git*}"
$ echo $GIT_DIR
/tmp/wallet-world-0/.git/worktrees/wallet-world-1
$ echo $PRJ_ROOT
/tmp/wallet-world-0
$ ls $PRJ_ROOT
some_file
$ if [[ -e $PRJ_ROOT/some_new_file ]] ; then echo "on new commit"; else echo "on initial commit"; fi
on initial commit
$ unset GIT_DIR
```

I still think you need `--show-toplevel` which in this example would get you;

```console
$ cd /tmp/wallet-world-0
$ git rev-parse --show-toplevel
/tmp/wallet-world-0
$ cd /tmp/wallet-world-1
$ git rev-parse --show-toplevel
/tmp/wallet-world-1
```

**SOMETHING INTERESTING I DID NOTICE WHILE GENERATING ABOVE** is that setting `GIT_DIR` in the environment will override the behaviour of `git rev-parse --absolute-git-dir` and force it to use whatever is in that env var. I mention this since, while you don't export `GIT_DIR` in the script, it might cause some confusion if you've been experimenting with this stuff.

Hope that's helpful and that I haven't just added more confusion! 



<details>
<summary>More context</summary>


So if a standard checkout is currently on <some_commit> you'd have a file structure like;

```
/git/wallet-world-0/         
/git/wallet-world-0/.git      - git state directory
/git/wallet-world-0/some_file - file in the commit
```

assuming you want to work in this working directory you want `PRJ_ROOT` to be `/git/wallet-world-0` right?

Now say you want a new worktree on a different commit, and assuming you run `git worktree add /git/wallet-world-1 <some_new_commit>`you get this created;

```
/git/wallet-world-1/         
/git/wallet-world-1/.git           - git state. links to /git/wallet-world-0/.git/worktrees/wallet-world-1
/git/wallet-world-1/some_file      - file in the commit
/git/wallet-world-1/some_new_file  - some file exclusive to <some_commit>
```

If you want to work on this worktree on `<some_new_commit>` you presumably want `PRJ_ROOT` to be `/git/wallet-world-1`?

`git rev-parse --absolute-git-dir` should always give you the state directory, but for worktrees this state directory is not in the same place as the checkout. So it'd return something like `/git/wallet-world-0/.git/`.

But what I think you want is not the state directory but the actually checked out contents, i.e. `/git/wallet-world-1/`.


I guess as another example, say I have a script that checks which commit we're on;

```sh
if [[ -e $PRJ_ROOT/some_new_file ]] ; then
    echo "on some_new_commit"
else
    echo "on some_commit"
fi
```

when we run the script in `wallet-world-0` we should expect the output `on some_commit` and when we run it in `wallet-world-1` we should expect `on some_new_commit`. 

I think using the new logic from 3ea5e756ca4c228aeb8586eb1, if we run `git rev-parse --absolute-git-dir` in `wallet-world-1` we'll get the output `/git/wallet-world-0/.git/worktrees/wallet-world-1`.
If we then use the expansion `${GIT_DIR%/.git*}` we'll end up with `/git/wallet-world-0/` so even though we're running in `wallet-world-1` `PRJ_ROOT` is going to point at `wallet-world-0`. 


</details>

File: contrib/direnv
Comment (cscutcher) [inline contrib/direnv:24]: Just in case the comment is missed in a resolved thread, I think this might still be a problem. See https://github.com/numtide/prj-spec/pull/4#discussion_r1192970170

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: One question:

Do we want:
- `PRJ_ROOT` point to worktree checkout; yet
- `PRJ_CONFIG_HOME` & al point to origin checkout (where `./.config` & al already live)

?

cc @zimbatm

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:24]: Yes indeed a problem remains. Ack!

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: Given that, with `PRJ_ID`, we have an escape hatch to lift the state into its XDG equivalents, it captures the intent of global shared state for reason XYZ.

With worktrees, users aren't hindered to do so as well.

Hence, you are right. We shall consider them separately.

**Comment by zimbatm** [comment]
@blaggacao happy to merge this once you tell me it's ready

**Comment by blaggacao** [comment]
@zimbatm I'll bump downstream to the tip.of this branch and keep running it for a while in real life. It looks like everything is addressed, now, so if that remains true I'll ping you in a couple of weeks. Thank you!

**Review by cscutcher — CHANGES_REQUESTED** [review]
Hopefully, this doesn't cause more problems. I only just noticed the review request and wanted to make sure the lack of change to `find_prj_root_with_git` was intentional.

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv:35]: I'm going to keep it due to "code parallelism".

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: 440705abeb8de9e63c703724a21802c89f6600c7

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: 440705abeb8de9e63c703724a21802c89f6600c7

File: contrib/direnv
Comment (blaggacao) [reply contrib/direnv]: 440705abeb8de9e63c703724a21802c89f6600c7

**Review by cscutcher — APPROVED** [review]

**Comment by blaggacao** [comment]
@zimbatm 9b0ffcd0fddcb261bcd73ad9dad18a096760b4a0 has already propagated for a while ot some real world environments via [this facility](https://github.com/paisano-nix/direnv/blob/bd36f51b9d566d8a36e230dab638abf3074625c3/lib#L24).

@cscutcher Has been an excellent reviewer. I greatly appreciate your engagement. It would not have been possible to polish this off so fast without your involvement.

All this to say: _This is reasonably ready for merge._ Should something arise, a fix it shall be.

**Pull request merged**
